bridge/jira/export.go

```diff
@@ -0,0 +1,449 @@
+			out <- core.NewExportNothing(
+				op.Id(), fmt.Sprintf(
+					"missing operation author credentials for user %.8s",
+					author.Id().String()))
+			continue
+		}
+
+		var id string
+		var exportTime time.Time
+		switch opr := op.(type) {
+		case *bug.AddCommentOperation:
+			comment, err := client.AddComment(bugJiraID, opr.Message)
+			if err != nil {
+				err := errors.Wrap(err, "adding comment")
+				out <- core.NewExportError(err, b.Id())
+				return
+			}
+			id = comment.ID
+			out <- core.NewExportComment(op.Id())
+
+			// cache comment id
+			self.cachedOperationIDs[op.Id()] = id
+
+		case *bug.EditCommentOperation:
+			if opr.Target == createOp.Id() {
+				// An EditCommentOpreation with the Target set to the create operation
+				// encodes a modification to the long-description/summary.
+				exportTime, err = client.UpdateIssueBody(bugJiraID, opr.Message)
+				if err != nil {
+					err := errors.Wrap(err, "editing issue")
+					out <- core.NewExportError(err, b.Id())
+					return
+				}
+				out <- core.NewExportCommentEdition(op.Id())
+				id = bugJiraID
+			} else {
+				// Otherwise it's an edit to an actual comment. A comment cannot be
+				// edited before it was created, so it must be the case that we have
+				// already observed and cached the AddCommentOperation.
+				commentID, ok := self.cachedOperationIDs[opr.Target]
+				if !ok {
+					// Since an edit has to come after the creation, we expect we would
+					// have cached the creation id.
+					panic("unexpected error: comment id not found")
+				}
+				comment, err := client.UpdateComment(bugJiraID, commentID, opr.Message)
+				if err != nil {
+					err := errors.Wrap(err, "editing comment")
+					out <- core.NewExportError(err, b.Id())
+					return
+				}
+				out <- core.NewExportCommentEdition(op.Id())
+				// JIRA doesn't track all comment edits, they will only tell us about
+				// the most recent one. We must invent a consistent id for the operation
+				// so we use the comment ID plus the timestamp of the update, as
+				// reported by JIRA. Note that this must be consistent with the importer
+				// during ensureComment()
+				id = fmt.Sprintf("%s-%d", comment.ID, comment.Updated.Unix())
+			}
+
+		case *bug.SetStatusOperation:
+			jiraStatus, hasStatus := statusMap[opr.Status.String()]
+			if hasStatus {
+				exportTime, err = UpdateIssueStatus(client, bugJiraID, jiraStatus)
+				if err != nil {
+					err := errors.Wrap(err, "editing status")
+					out <- core.NewExportError(err, b.Id())
+					// Failure to update status isn't necessarily a big error. It's
+					// possible that we just don't have enough information to make that
+					// update. In this case, just don't export the operation.
+					continue
+				}
+				out <- core.NewExportStatusChange(op.Id())
+				// TODO(josh)[c2c6767]: query changelog to get the changelog-id so that
+				// we don't re-import the same change.
+				id = bugJiraID
+			} else {
+				out <- core.NewExportNothing(
```

Comment:
I would think that if the bridge is unable to export a status, that would be an error

Reply:
Although I didn't envision to have bridge return multiple error events, it's entirely possible. It'd be nice if we can stay within a single event per bug but if not that's ok.

Reply:
haha, I see we went to the same direction. Let's do the warning then.

CC @A-Hilaly

Reply:
Note to self: Also do this for un-implemented changelog events.

Reply:
This is done.